File: src/3d/render_ray.c

#include "render_ray.h"

#define MAX(a, b) ({ \
    __typeof__(a) _a = (a); \
    __typeof__(b) _b = (b); \
    _a > _b ? _a : _b; \
})

struct Pixel raycolor(struct Scene* scene, struct Ray* r, struct Interval* interval) {
    struct Record info = {0};
    struct Pixel color = make_pixel(0, 0, 0);
    if(scene->objects.intersect(&scene->objects, r, interval, &info)) {
        struct Point3F32 tmp = r->point_at(r, info.time);
        struct Vec3 lvec = point3F32_sub(&scene->camera->eye, &tmp);
        lvec = vec3_normalized(&lvec);

        float lamfact = MAX(0.0f, vec3_dot(&lvec, &info.normal));
        color = make_pixel(
            info.color.r * lamfact + info.color.r * scene->ambient,
            info.color.g * lamfact + info.color.g * scene->ambient,
            info.color.b * lamfact + info.color.b * scene->ambient
        );
    }
    else {
        color = scene->background;
    }
    return color;
}

void raytrace(struct Scene* scene, struct Image* img) {
    scene->camera->set_resolution(scene->camera, img->width, img->height);

    for(uint16_t j=0; j < img->height; j++) {
        for(uint16_t i=0; i < img->width; i++) {
            struct Ray ray = scene->camera->ij_ray(scene->camera, i, j);
            struct Interval interval = make_interval(0.00, INFINITY);
            struct Pixel color = raycolor(scene, &ray, &interval);
            color = color.quantize(&color, 255);
            img->set_pixel(img, (struct Point2I16){i, j}, color);
        }
    }
}

File: src/3d/render_ray.zig
const std = @import("std");
const scenedef = @import("scenedef.zig");
const image = @import("image.zig");
const ray3d = @import("ray3d.zig");
const gfx_types = @import("gfx_types.zig");
const models = @import("models.zig");

fn raycolor(scene: *scenedef.Scene, r: *const ray3d.Ray, interval: *ray3d.Interval, nreflect: u8) gfx_types.Pixel {
    var info: models.Record = undefined;
    var color = gfx_types.Pixel{ .r = 0, .g = 0, .b = 0 };

    const hit = scene.objects.intersect(r, interval, &info);
    if (hit) {
        const k = info.color;
        const hitpt = info.point;
        const n = info.normal;

        color = k.ambient.mul(scene.ambient);

        for (scene.lights.items) |light| {
            var lpos: gfx_types.Point(f32, 3) = undefined;
            var lcolor: gfx_types.Pixel = undefined;
            lpos, lcolor = light;

            var lvec = lpos.sub(&hitpt);

            var shadow_interval = ray3d.Interval{ .low = 0.0001, .high = 1 };
            var shadow_rec: models.Record = undefined;
            if (!(scene.shadows and scene.objects.intersect(&.{ .start = hitpt, .dir = lvec }, &shadow_interval, &shadow_rec))) {
                lvec = lvec.normalized();
                const lambert = @max(0.0, lvec.dot(&n));
                const diffuse = k.diffuse.mul(lambert);
                color = color.add(&diffuse);

                var new_color = color.add(&k.specular.mul(lcolor));

                const vvec = r.dir.normalized().neg();
                const h = lvec.add(&vvec).normalized();
                const shine = std.math.pow(f32, @max(0.0, h.dot(&n)), k.shininess);

                new_color = new_color.mul(shine);
                color = color.add(&new_color);
            }
        }
        if (k.reflect != null and nreflect > 0) {
            var ray_interval = ray3d.Interval{ .low = 0.0001, .high = std.math.inf(f32) };
            var rcolor = raycolor(scene, &.{ .start = hitpt, .dir = r.dir.reflect(&n) }, &ray_interval, nreflect - 1);
            rcolor = rcolor.mul(k.reflect.?);
            color = color.add(&rcolor);
        }
    } else {
        color = scene.background;
    }

    return color;
}

pub fn raytrace(scene: *scenedef.Scene, img: *image.Image) void {
    scene.camera.set_resolution(img.width, img.height);

    for (0..img.height) |j| {
        for (0..img.width) |i| {
            var ray = scene.camera.ij_ray(@floatFromInt(i), @floatFromInt(j));
            var interval = ray3d.Interval{ .low = 0.00, .high = std.math.inf(f32) };
            var color = raycolor(scene, &ray, &interval, scene.reflections);
            color = color.quantize(255);
            img.set_pixel(.{ .vals = .{ @intCast(i), @intCast(j) } }, color);
        }
    }
}
